### PR TITLE
New suggestions System

### PR DESCRIPTION
**Pull request by Speiger**
New Suggestion System allows for a much nicer way to implement/use suggestions.

### CODE REVIEW

File: src/main/java/carbonconfiglib/api/ISuggestionProvider.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	public static class SingleSuggestion implements ISuggestionProvider {
+		Suggestion value;
+		
+		private SingleSuggestion(Suggestion value) {
+			this.value = value;
+		}
+		
+		@Override
+		public void provideSuggestions(Consumer<Suggestion> output, Predicate<Suggestion> filter) {
+			if(filter.test(value)) output.accept(value);
```
Comment (Meduris) [inline src/main/java/carbonconfiglib/api/ISuggestionProvider.java:43]: I am not necessarily a fan of writing it in one line like this

File: src/main/java/carbonconfiglib/api/ISuggestionProvider.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+		List<Suggestion> suggestions;
+		
+		private SimpleSuggestion(List<Suggestion> suggestions) {
+			this.suggestions = suggestions;
+		}
+		
+		@Override
+		public void provideSuggestions(Consumer<Suggestion> output, Predicate<Suggestion> filter) {
+			for(Suggestion value : suggestions) {
+				if(filter.test(value)) output.accept(value);
```
Comment (Meduris) [inline src/main/java/carbonconfiglib/api/ISuggestionProvider.java:57]: I am not necessarily a fan of writing it in one line like this

File: src/main/java/carbonconfiglib/api/ISuggestionProvider.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+	public static class Suggestion {
+		String name;
+		String value;
+		Object type;
+		
+		private Suggestion(String name, String value, Object type) {
+			this.name = name;
+			this.value = Objects.requireNonNull(value);
+			this.type = type;
+			if(value.trim().isEmpty()) throw new IllegalArgumentException("Value isn't allowed to be empty");
```
Comment (Meduris) [inline src/main/java/carbonconfiglib/api/ISuggestionProvider.java:91]: I am not necessarily a fan of writing it in one line like this

**Review by Meduris — APPROVED** [review]
I mentioned it in a few places, I don't really like the having the executed code on the same line as the if statements, but that is nothing too mayor

**Comment by Speiger** [comment]
@Meduris yeah the single line was a compromise on my side.
Because the other way around would have been.
```java
@Override
public void provideSuggestions(Consumer<Suggestion> output, Predicate<Suggestion> filter) {
    if(filter.test(value)) {
        output.accept(value);
    }
}
````
And i thought the single line solution would have been the compromise.

Also its kinda already the standard.

**Pull request merged**